text/0000-fusion-release-process.md

```diff
@@ -0,0 +1,71 @@
+
+### Private package support
+
+Due to containerization, this project should be able to support both open source and closed source projects given that there is sufficient infrastructure. For open source projects we plan on running this within Buildkite, verifying each package on one or more parallel workers. A list of additional private projects may be contained within a gitignored file, or pulled down at runtime from another source. All of the existing topological sorting and package linking should continue to work.
+
+### Release verification
+
+We will verify releases by running unit tests, linters, flow and integration tests across the HEAD of all packages. Only by running on HEAD can we discern whether or not we are safe to release or not.
+
+### Release lockstep publishing
+
+Assuming that all tests pass, we can kick off a pipeline job which will use our built monorepo to publish to the NPM registry. We can leverage a Buildkite input or select dropdown to choose the next version. (https://buildkite.com/docs/pipelines/block-step#select-input-attributes)
+
+Even though we did not bootstrap with Lerna, we can take advantage of their publishing scripts to lockstep-publish all FusionJS repositories. If we do publish from this pipeline, we would need the repository to commit both package.json and yarn.lock to each repository on Github.
+
+# Drawbacks
+
+### Complexity
+
+There's a fair amount of complexity involved in our current fusion-release repository, and it will likely grow over time. Pieces that add complexity include usage of a topological sorting library, as well as a pipeline which defines the entire testing pipeline. This pipeline is currently not generated due to slight differences between repos, and will need to be updated as we add additional plugins and test suites. One alternative is creating a yarn script for each plugin which will allow us to shell out to that command per repository. E.g., yarn run release-verification.
+
+### Machine cost
+
+Bundling FusionJS as a monorepo results in a 5GB docker image along with dozens of hours of machine time. Compared to engineering cost the price is miniscule, but there is a cost to running all of these compute hours on AWS (50 machine hours being $1 per commit to test on spot instances or so). We would ideally receive a budget for something like "500 compute hours per pull request" and not have to worry about cost at all.
+
+# Alternatives
+
+* Keep doing what we're doing today.
+
+# Adoption strategy
+
+### Release version lockstepping
+
+We would stop publishing packages individually, and instead lockstep the entire release. This would likely save developers time as they would not need to manually version bump.
```

Comment:
Are you proposing publishing all code in a single package (as opposed to separate packages), or merely versioning all packages in lockstep?